core/keyboard_nav/navigation.js

```diff
@@ -723,108 +859,62 @@ Blockly.navigation.error = function(msg) {
-  } else if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_FLYOUT) {
-    Blockly.navigation.selectNextBlockInFlyout();
-  } else if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_TOOLBOX) {
-    Blockly.navigation.nextCategory();
-  }
-});
+Blockly.navigation.ACTION_NEXT = new Blockly.Action(
+    Blockly.navigation.actionNames.NEXT, 'Goes to the next location');
 
 /**
- * The action to go in.
+ * The in action.
  * @type Blockly.Action
  */
-Blockly.navigation.ACTION_IN = new Blockly.Action('in', 'Goes in', function() {
-  if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_WS) {
-    Blockly.navigation.cursor_.in();
-  } else if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_TOOLBOX) {
-    Blockly.navigation.inCategory();
-  }
-});
+Blockly.navigation.ACTION_IN = new Blockly.Action(
+    Blockly.navigation.actionNames.IN, 'Goes in');
 
 /**
  * The action to try to insert a block.
  * @type Blockly.Action
  */
-Blockly.navigation.ACTION_INSERT = new Blockly.Action('insert',
-    'Tries to connect the current location to the marked location', function() {
-      if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_WS) {
-        Blockly.navigation.modify();
-      }
-    });
+Blockly.navigation.ACTION_INSERT = new Blockly.Action(
+    Blockly.navigation.actionNames.INSERT,
+    'Tries to connect the current location to the marked location');
 
 /**
  * The action to mark a certain location.
  * @type Blockly.Action
  */
-Blockly.navigation.ACTION_MARK = new Blockly.Action('mark', 'Marks the current location', function() {
-  if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_WS) {
-    Blockly.navigation.handleEnterForWS();
-  } else if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_FLYOUT) {
-    Blockly.navigation.insertFromFlyout();
-  }
-});
+Blockly.navigation.ACTION_MARK = new Blockly.Action(
+    Blockly.navigation.actionNames.MARK, 'Marks the current location');
 
 /**
  * The action to disconnect a block.
  * @type Blockly.Action
  */
-Blockly.navigation.ACTION_DISCONNECT = new Blockly.Action('disconnect', 'Disconnect the blocks', function() {
-  if (Blockly.navigation.currentState_ === Blockly.navigation.STATE_WS) {
-    Blockly.navigation.disconnectBlocks();
-  }
-});
+Blockly.navigation.ACTION_DISCONNECT = new Blockly.Action(
+    Blockly.navigation.actionNames.DISCONNECT, 'Disconnect the blocks');
 
 /**
  * The action to open the toolbox.
  * @type Blockly.Action
  */
-Blockly.navigation.ACTION_TOOLBOX = new Blockly.Action('toolbox', 'Open the toolbox', function() {
-  if (!Blockly.getMainWorkspace().getToolbox()) {
-    Blockly.navigation.focusFlyout();
-  } else {
-    Blockly.navigation.focusToolbox();
-  }
-});
+Blockly.navigation.ACTION_TOOLBOX = new Blockly.Action(
+    Blockly.navigation.actionNames.TOOLBOX, 'Open the toolbox');
 
 /**
  * The action to exit the toolbox or flyout.
  * @type Blockly.Action
```

Comment:
* @type {Blockly.Action}

Reply:
I would actually go for !Blockly.Action here. since you call onBlocklyAction in blockly.js with Blockly.navigation.ACTION_EXIT and onBlocklyAction only accepts non nullable actions.